parser/parser.c3

```diff
@@ -0,0 +1,1381 @@
+
+        let name : ProgramSlice = cursor:get_next_ascii_only_identifier();
+        variable_access = {name};
+        return {4, ref_to_addr(&variable_access), get_meta(name)};
+    }
+
+    syntax_error(&cursor, sv("unexpected symbol '%', cannot start an expression"), decode_rune(cursor:get_char()));
+    return {0, 0, meta(0, 0)};
+}
+
+function parse_maybe_unary_or_borrow_expression : (cursor : Cursor&) -> Expression = {
+    let start_offset : isize = cursor.position;
+
+    if is_one_of(cursor:get_char(), "!"[0], "-"[0], "+"[0], "~"[0]) {
+        let name : ProgramSlice = cursor:slice(1);
+        cursor.position += 1;
+        cursor:try_skip_white_space();
+
+        let rhs : Expression = cursor:parse_maybe_unary_or_borrow_expression();
+
+        let operator_call : UnaryOperator& = &cursor.allocator:allocate<UnaryOperator>();
+        operator_call = {name, rhs, meta(start_offset, rhs.meta.end_offset)};
+
+        return {9, ref_to_addr(&operator_call), operator_call.meta};
+    }
+
+    if cursor:try_skip_next("&") {
+        let is_const : bool = cursor:try_skip_next_and_surrounding_white_space("const");
+        let rhs : Expression = cursor:parse_maybe_unary_or_borrow_expression();
+
+        let borrow : Borrow& = &cursor.allocator:allocate<Borrow>();
+        borrow = {is_const, rhs, meta(start_offset, borrow.meta.end_offset)};
+        return {11, ref_to_addr(&borrow), borrow.meta};
+    }
+
+    return cursor:parse_non_operator_expression();
+}
+
+function [T] parse_next_operator_subexpression : (cursor: Cursor&, op_iter : T) -> Expression = {
+    if op_iter:has_next() {
+        return cursor:parse_expression_with_order_of_operator(op_iter);
+    } else {
+        return cursor:parse_maybe_unary_or_borrow_expression();
+    }
+}
+
+function [T] parse_expression_with_order_of_operator : (cursor : Cursor&, op_iter : T) -> Expression = {
+    // TODO: make the pattern matching robust enough to put this in the standard library
+    let current_op : StringView = op_iter:__builtin_get_next();
+    let lhs : Expression = cursor:parse_next_operator_subexpression(op_iter);
+
+    let last_operator_start : isize = cursor.position;
+    while cursor:try_skip_next_operator(current_op) {
+        let op_name : ProgramSlice = {last_operator_start, cursor:skip_ws_return_offset()};
+
+        let rhs : Expression = cursor:parse_next_operator_subexpression(op_iter);
+
+        let meta : Meta = meta(lhs.meta.start_offset, cursor.position);
+        let operator_call : Operator& = &cursor.allocator:allocate<Operator>();
+        operator_call = {op_name, lhs, rhs, meta};
+
+        lhs = {8, ref_to_addr(&operator_call), meta};
+        last_operator_start = cursor.position;
+    }
+    return lhs;
+}
+
+function parse_expression : (cursor : Cursor&) -> Expression = {
+    // TODO: this size argument needs to somehow be automatic
+    // TODO: all items on the same line should have the same priority
```

Comment:
There are algorithms for this https://en.wikipedia.org/wiki/Operator-precedence_parser#Precedence_climbing_method